### PR TITLE
Adding useful utility methods to clip overlapping bases for mapped

### PR DESCRIPTION
**Pull request by nh13**
pairs.

### CODE REVIEW

**Comment by nh13** [comment]
Moved utility methods from https://github.com/broadinstitute/picard/pull/332 into htsjdk.

**Comment by tfenne** [comment]
:+1:  from me.

File: src/tests/java/htsjdk/samtools/SAMUtilsTest.java [hunk 1]
```diff
@@ -42,4 +44,35 @@ public void testCompareMapqs() {
...
+        record.setReadBases("AAAAAAAAAA".getBytes());
+
+        final int numToClip = SAMUtils.getNumOverlappingBasesToClip(record);
+        Assert.assertEquals(numToClip, 1);
+
+        SAMUtils.clipOverlappingBases(record, numToClip, false); // Side-effects are OK
+
+        final Cigar cigar = record.getCigar();
+        Assert.assertEquals(cigar.numCigarElements(), 2);
+        Assert.assertEquals(cigar.getCigarElement(0).getLength(), 4); // 4M
```
Comment (yfarjoun) [inline src/tests/java/htsjdk/samtools/SAMUtilsTest.java]: Any reason not to compare the cigars directly? Or the cigar-strings if that doesn't work?

File: src/java/htsjdk/samtools/SAMUtils.java [hunk 2]
```diff
@@ -970,4 +971,77 @@ public static String getCanonicalRecordName(final SAMRecord record) {
...
+     */
+    public static int getNumOverlappingBasesToClip(final SAMRecord rec) {
+        // NB: ignores how to handle supplemental records when present for both ends by just using the mate information in the record.
+
+        if (!rec.getReadPairedFlag() || rec.getReadUnmappedFlag() || rec.getMateUnmappedFlag()) return 0;
+
+        // Only clip records that are left-most in genomic order and overlapping.
+        if (rec.getMateAlignmentStart() < rec.getAlignmentStart()) return 0; // right-most, so ignore.
+
+        final int numBasesToClip = rec.getAlignmentEnd() - rec.getMateAlignmentStart()  + 1;
```
Comment (yfarjoun) [inline src/java/htsjdk/samtools/SAMUtils.java]: hmmm. I think that there's a mistake here. 

The overlap is based on reference bases, but the clipping is done on read bases. So if there's an insertion or deletion in the overlap region, I suspect that the calculation will be wrong.

Could you add a few more test cases that have insertions and/or deletions in the overlap region?

Comment (tfenne) [inline src/java/htsjdk/samtools/SAMUtils.java]: Doh - agreed.  It looks like an older version of this method crept back in.

**Comment by nh13** [comment]
@yfarjoun @tfenne back to you, including some new test cases.

File: src/tests/java/htsjdk/samtools/SAMUtilsTest.java [hunk 3]
```diff
@@ -42,4 +44,125 @@ public void testCompareMapqs() {
...
+        record.setMateReferenceIndex(0);
+        record.setReadBases("AAAAAAAAAA".getBytes());
+
+        final int numToClip = SAMUtils.getNumOverlappingBasesToClip(record);
+        Assert.assertEquals(numToClip, 1);
+
+        SAMUtils.clipOverlappingBases(record, numToClip, false); // Side-effects are OK
+
+        final Cigar cigar = record.getCigar();
+        final Cigar expectedCigar = new Cigar(Arrays.asList(new CigarElement(4, CigarOperator.M), new CigarElement(6, CigarOperator.S))); // 4M6S
```
Comment (yfarjoun) [inline src/tests/java/htsjdk/samtools/SAMUtilsTest.java]: any reason not to use TextCigarCodec.decode ?

**Comment by nh13** [comment]
@yfarjoun thanks for the tip on `TextCigarCodec`.

**Comment by nh13** [comment]
@yfarjoun renaming the functions as we really are computing the number of aligned bases clipped.

File: src/tests/java/htsjdk/samtools/SAMUtilsTest.java [hunk 4]
```diff
@@ -42,4 +44,113 @@ public void testCompareMapqs() {
...
+         * Tests that if we need to clip a read with an deletion that overlaps
+         */
+
+        // setup the record
+        final SAMFileHeader header = new SAMFileHeader();
+        header.addSequence(new SAMSequenceRecord("1", 1000));
+        final SAMRecord record = new SAMRecord(header);
+        record.setReadPairedFlag(true);
+        record.setCigar(
+                new Cigar(Arrays.asList(
```
Comment (yfarjoun) [inline src/tests/java/htsjdk/samtools/SAMUtilsTest.java]: one more :-)

**Comment by nh13** [comment]
@yfarjoun  thanks for taking the time to review all of our pull requests, I know it is time consuming to be thorough, thank-you!

**Comment by yfarjoun** [comment]
You're welcome. Once I find time to respond to comment in my own pull
request, I'd appreciate a thorough review as well :-)

**Pull request merged**